nixos/modules/services/networking/wireguard.nix

```diff
@@ -0,0 +1,237 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+
+  cfg = config.networking.wireguard;
+
+  kernel = config.boot.kernelPackages;
+ 
+  # interface options
+
+  interfaceOpts = { name, ... }: {
+
+    options = {
+
+      ips = mkOption {
+        example = [ "192.168.2.1/24" ];
+        default = [];
+        type = with types; listOf str;
+        description = "The IP addresses of the interface.";
+      };
+
+      privateKey = mkOption {
+        example = "yAnz5TF+lXXJte14tji3zlMNq+hd2rYUIgJBgB3fBmk=";
+        type = types.str;
+        description = "Base64 private key generated by wg genkey.";
+      };
+
+      presharedKey = mkOption {
+        default = null;
+        example = "rVXs/Ni9tu3oDBLS4hOyAUAa1qTWVA3loR8eL20os3I=";
+        type = with types; nullOr str;
+        description = ''base64 preshared key generated by wg genpsk. Optional,
+        and may be omitted. This option adds an additional layer of
+        symmetric-key cryptography to be mixed into the already existing
+        public-key  cryptography, for post-quantum resistance.'';
+      };
+
+      listenPort = mkOption {
+        default = null;
+        type = with types; nullOr int;
+        example = 51820;
+        description = ''16-bit port for listening. Optional; if not specified,
+        automatically generated based on interface name.'';
+      };
+
+      preSetup = mkOption {
+        example = literalExample ''
+          ${pkgs.iproute}/bin/ip netns add foo
+        '';
+        default = "";
+        type = types.lines;
+        description = ''A set of commands called at the start of the interface
+        setup.'';
+      };
+
+      postSetup = mkOption {
+        example = literalExample ''
+          printf 'nameserver 10.200.100.1' | ${pkgs.openresolv}/bin/resolvconf -a wg0 -m 0
```

Comment:
this example will not work, because it is not executed within in a shell - take a look at my version.